torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp

```diff
@@ -1,372 +1,904 @@
+          }),
+      consumer_edges.end());
+
+  return consumer_edges;
+}
+
+// Assuming sg1 and sg2 are connected, figure out which is the consumer
+const ExprGroup* getProducer(const ExprGroup* sg1, const ExprGroup* sg2) {
+  for (auto producer_edge : sg1->producer_edges_) {
+    if (producer_edge->from_ == sg2) {
+      return sg2;
     }
   }
-  reorderSegmentBreadthFirst(seg_begin, seg_end);
+
+  for (auto consumer_edge : sg1->consumer_edges_) {
+    if (consumer_edge->to_ == sg2) {
+      return sg1;
+    }
+  }
+
+  return nullptr;
 }
 
-void mergeNonRootGroupsIntoRootGroups(
-    TargetGroupMap& computed_at_exprs,
-    ExprTargetMap& target_map) {
-  for (auto it = computed_at_exprs.begin(); it != computed_at_exprs.end();) {
-    TensorView* target = it->first;
-    if (target->hasComputeAt()) {
-      Expr* target_expr = target->definition();
-      TensorView* target_of_target = target_map.at(target_expr);
-      auto& target_group = computed_at_exprs.at(target_of_target);
-      auto pos =
-          std::find(target_group.begin(), target_group.end(), target_expr);
-      TORCH_INTERNAL_ASSERT(pos != target_group.end());
-      target_group.insert(pos, it->second.begin(), it->second.end());
-      // Update the target map
-      for (auto& inserted_expr : it->second) {
-        TORCH_INTERNAL_ASSERT(target_map.at(inserted_expr) == target);
-        target_map.at(inserted_expr) = target_of_target;
-      }
-      it = computed_at_exprs.erase(it);
+} // namespace
+
+// Disconect group from neighbors, and return edges that were disconnected
+std::unordered_set<ExprGroupConnections*> ExprSegmentationSorter::
+    disconnectGroup(ExprGroup* group) {
+  std::unordered_set<ExprGroupConnections*> removed_edges(
+      group->producer_edges_.begin(), group->producer_edges_.end());
+
+  for (auto edge : group->producer_edges_) {
+    auto from = edge->from_;
+    auto& from_edges = from->consumer_edges_;
+    auto from_edge_it = std::find(from_edges.begin(), from_edges.end(), edge);
+    TORCH_INTERNAL_ASSERT(
+        from_edge_it != from_edges.end(), "Could not find edge to remove.");
+    from_edges.erase(from_edge_it);
+  }
+
+  for (auto edge : group->consumer_edges_) {
+    auto to = edge->to_;
+    auto& to_edges = to->producer_edges_;
+    auto to_edge_it = std::find(to_edges.begin(), to_edges.end(), edge);
+    TORCH_INTERNAL_ASSERT(
+        to_edge_it != to_edges.end(), "Could not find edge to remove.");
+    to_edges.erase(to_edge_it);
+  }
+
+  group->producer_edges_.clear();
+  group->consumer_edges_.clear();
+
+  return removed_edges;
+}
+
+ExprGroup* ExprSegmentationSorter::makeMergedNode(
+    ExprGroup* sg1,
+    ExprGroup* sg2) {
+  std::vector<IterDomain*> resulting_ca_axes;
+  auto& domain1 = sg1->payload()->ca_domains;
+  auto& domain2 = sg2->payload()->ca_domains;
+  auto it1 = domain1.begin();
+  auto it2 = domain2.begin();
+
+  while (it1 != domain1.end() && it2 != domain2.end()) {
```

Comment:
Shouldn't this use `||` instead of `&&`? With `&&`, the first two if clauses are never used.

Reply:
Can we just do something like:

```
auto resulting_ca_axes = domain1;
for (auto id2 : domain2) {
  if (std::none_of(domain1.begin(), domain1.end(), [&](auto id1) {
         return caIndexMap().areMapped(id1, id2);
      })) {
      resulting_ca_axes.push_back(id2);
  }
}
```

Reply:
We need all iteration domains from both to propagate, and we need to maintain relative order while trying to reduce duplicates when we can. I added some more comments.

Reply:
You were right about the `||` instead of `&&`, changed that as well.